Sarif reporting - temp 2

de-jcup opened the pull request:
Just to identify current changes between main and sarif-reporting branch

de-jcup committed:
Feature 6875 provide sarif reports (#2)

- inital SARIF integration
- contains applied review changes from origin PR wanted by ZAP team

de-jcup committed:
Report support now created inside report template directly

- ReportHelper changes reverted
- Creating support object now no longer via helper method but
  directly inside template

de-jcup committed:
Drop "evidence" from property bag but use region snippet instead

- we do not need evidence in property bag. region/snippet is much
  better
- region/snippet changed from owasp "attack" content to evidence
  content. Also dropped former fallback mechanism.

de-jcup committed:
Moved sarif repporting integration test + fixed test

- sarif reporting integration test is now in dedicated sarif
  sub package for testing (like other sarif parts)
- evidence was removed from property bag, so test did fail.
  fixed this

de-jcup committed:
Moved test alert and alert node builders to sarif sub folder

de-jcup committed:
Shrinking webresponse and webrequest body when necessary

- currently 8k webrequest or 32k webresponse are allowed before
  shrinking.

de-jcup committed:
Fixed failing test

de-jcup committed:
Added javadoc + applied spotless formattter

de-jcup committed:
Merge branch 'main' into sarif-reporting

de-jcup committed:
Handle application/javascript and pendants as text not binary

de-jcup committed:
Improved sarif report shrinking for body and evidence snippets

Jeeppler commented on addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBigContentShrinker.java:
```diff
@@ -0,0 +1,158 @@
+        int charsAfter = charsBefore;
+
+        /* calculate beginning */
+        boolean shrinkMarketAtBeginning = true;
```
```suggestion
        boolean shrinkMarkerAtBeginning = true;
```

Jeeppler commented on addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifBinaryContentDetector.java:
```diff
@@ -0,0 +1,60 @@
+            // if not set, we assume it is binary
+            return true;
+        }
+        return !(contentTypeLowerCased.startsWith("text")
```
This is hard to read and evaluate as many things are going on here. I would prefer the use of `if` and `else`, even if this increases the verbosity.

Jeeppler commented on addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifReportDataSupport.java:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    private List<SarifResult> createResults() {
+        List<SarifResult> results = new ArrayList<>();
```
I would prefer a linked list rather than a ArrayList. The reason, being that for a array list to grow beyond it's initial capacity, a memory copy needs to happen (memory reallocation). A linked list instead can simply grow.

Or in this case move the statement below `allAlerts` and initialize the arraylist with the size of `allAlerts`.

de-jcup replied:
Linked lists have some caveats as well - the sorting done at the end will be much slower compared to array list.
But I will init the array list with the correct initial capacity - as you suggested.

Jeeppler commented on addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifReportDataSupport.java:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    private List<Alert> collectAllAlerts(AlertNode rootNode) {
+        List<Alert> list = new ArrayList<>();
```
I would prefer a linked list rather than a ArrayList. The reason, being that for a array list to grow beyond it's initial capacity, a memory copy needs to happen (memory reallocation). A linked list instead can simply grow.

Jeeppler commented on addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifResult.java:
```diff
@@ -0,0 +1,420 @@
+            /* ------------ */
+            String shortEvidenceSnippet =
+                    bigContentShrinker.shrinkTextWithoutMarkers(
+                            alert.getEvidence(), MAX_ALLOWED_EVICENCE_SNIPPET_SIZE);
```
```suggestion
                            alert.getEvidence(), MAX_ALLOWED_EVIDENCE_SNIPPET_SIZE);
```

Jeeppler commented on addOns/reports/src/main/java/org/zaproxy/addon/reports/sarif/SarifToolData.java:
```diff
@@ -0,0 +1,201 @@
+                    .setTaxonomyInformationUri("https://www.zaproxy.org/")
+                    .build();
+
+    private static final SarifToolDataProvider CWE_WITH_4_4_TAXONOMY =
```
The CWE list is now at 4.7: https://cwe.mitre.org/data/downloads.html

Jeeppler commented on addOns/reports/src/test/java/org/zaproxy/addon/reports/sarif/SarifBase64EncoderTest.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @Test
+    void testdataResultsInCorrectBase64Value() {
```
Please add a test which contains an emoji. Emojis are not just ASCII symbols. Emojis are an easy way to test UTF-8/Unicode handling.

Jeeppler commented on addOns/reports/src/test/java/org/zaproxy/addon/reports/sarif/SarifBigContentShrinkerTest.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    // @formatter:off
+    @CsvSource({
```
What happens in case of `0` or `-2`?

Jeeppler commented on addOns/reports/src/test/java/org/zaproxy/addon/reports/sarif/SarifReportIntegrationTest.java:
```diff
@@ -0,0 +1,679 @@
+        assertEquals(expectedReferences, list);
+    }
+
+    private void assertRuleHasCweEntrAndGuidsNotDuplicated(
```
```suggestion
    private void assertRuleHasCweEntryAndGuidsNotDuplicated(
```

Jeeppler commented on addOns/reports/src/test/java/org/zaproxy/addon/reports/sarif/SarifReportIntegrationTest.java:
```diff
@@ -0,0 +1,679 @@
+            InspectionContext inspectionContext, SarifToolDataProvider cwe, JsonNode rule) {
+
+        // relationships
+        JsonNode relationShips = rule.get("relationships");
```
```suggestion
        JsonNode relationships = rule.get("relationships");
```

Jeeppler commented on addOns/reports/src/test/java/org/zaproxy/addon/reports/sarif/SarifReportIntegrationTest.java:
```diff
@@ -0,0 +1,679 @@
+        JsonNode relationShips = rule.get("relationships");
+        assertTrue(relationShips.isArray());
+        ArrayNode relationShipsArray = (ArrayNode) relationShips;
+        for (JsonNode relationShip : relationShipsArray) {
```
```suggestion
        for (JsonNode relationship : relationshipsArray) {
```

Jeeppler reviewed (changes requested):
See comments.

de-jcup committed:
Hide authorization header in SARIF reports

de-jcup committed:
Fix Java 8 compile problem

- String.isBlank() is not available in Java 8, so replaced with
  String.isEmpty() - this is enough here and works with Java 8

de-jcup committed:
Merge branch 'main' into sarif-reporting

de-jcup committed:
Merge branch 'sarif-reporting' into feature-hide-credentials-in-header

de-jcup committed:
Merge pull request #10 from mercedes-benz/feature-hide-credentials-in-header

Hide authorization header in SARIF reports

de-jcup committed:
Internal review changes

- fixed some typos
- array initialization with defined size
- used LinkedList for a growing list without index usage
- made some logic more readable
- switched to MITRE CWE 4.8 taxonomay, updated links etc.
